### PR TITLE
Update fast-classpath-scanner (2.8.1) to  classgraph (4.8.104)

### PR DESCRIPTION
**Pull request by NingZhuRedsilence**
Summary
==============================

`fast-classpath-scanner` has been deprecated; `classpath` is its successor. Replacing `fast-classpath-scanner` with `classgraph` caused a few code changes because the `classgraph` API has changed.

Additional Details
==============================

Before changing to `classpath`, with `fast-classpath-scanner` up to version `3.1.13`, unit tests written using `JenkinsPipelineSpecification` was throwing 
```
java.lang.OutOfMemoryError: GC Overhead Limit Exceeded
``` 
at [this line](https://github.com/ExpediaGroup/jenkins-spock/blob/935c4f2fc0d1fbe695cc6237597fc5a0cbe5e84c/src/main/groovy/com/homeaway/devtools/jenkins/testing/JenkinsPipelineSpecification.groovy#L1109).

Checklist
==============================

Testing
-------------------------

(Remove this checklist and replace it with "N/A - no code changes" if this PR does not modify source code)

* [x] I have manually verified that my code changes do the right thing.
* [x] I have run the tests and verified that my changes do not introduce any regressions.
* [ ] I have written unit tests to verify that my code changes do the right thing and to protect my code against regressions

Documentation
-------------------------

(Remove this checklist and replace it with "N/A - no code changes" if this PR does not modify source code)

* [x] I have updated the "Unreleased" section of `CHANGELOG.md` with a brief description of my changes.
* [ ] I have updated code comments - both GroovyDoc/JavaDoc-style comments and inline comments - where appropriate.
* [x] I have read `CONTRIBUTING.md` and have followed its guidance.

### CODE REVIEW

File: pom.xml [hunk 1]
```diff
@@ -276,10 +276,15 @@
 			<artifactId>reflections</artifactId>
 			<version>0.9.12</version>
 		</dependency>
+		<!--<dependency>-->
+		<!--<groupId>io.github.lukehutch</groupId>-->
+		<!--<artifactId>fast-classpath-scanner</artifactId>-->
+		<!--<version>3.1.13</version>-->
+		<!--</dependency>-->
```
Comment (awittha) [inline pom.xml]: No need to retain this as a comment; the git history will do that.

File: src/main/java/com/homeaway/devtools/jenkins/testing/WholeClasspathPipelineExtensionDetector.java [hunk 2]
```diff
@@ -77,10 +81,10 @@
...
 			} else {
 				LOG.warn(
-					"Failed to get some classes of type [{}] in package [{}]. For detailed error messages, set the system property PipelineExtensionDetector.expandFailures=true. Failures: [{}]",
-					_supertype,
-					_package,
-					failures.keySet() );
+						"Failed to get some classes of type [{}] in package [{}]. For detailed error messages, set the system property PipelineExtensionDetector.expandFailures=true. Failures: [{}]",
+						_supertype,
+						_package,
+						failures.keySet() );
```
Comment (awittha) [inline src/main/java/com/homeaway/devtools/jenkins/testing/WholeClasspathPipelineExtensionDetector.java]: Extra layer of indentation added here.

```suggestion
					"Failed to get some classes of type [{}] in package [{}]. For detailed error messages, set the system property PipelineExtensionDetector.expandFailures=true. Failures: [{}]",
					_supertype,
					_package,
					failures.keySet() );
```

File: src/main/java/com/homeaway/devtools/jenkins/testing/WholeClasspathPipelineExtensionDetector.java [hunk 3]
```diff
@@ -94,7 +98,11 @@
 
 		HashMap<String, Throwable> failures = new HashMap<>();
 
-		List<String> annotated_classnames = new FastClasspathScanner(_package.orElse("") ).scan().getNamesOfClassesWithAnnotation( _annotation );
+		List<String> annotated_classnames = new ClassGraph()
+				.enableAnnotationInfo()
+				.enableClassInfo()
+				.acceptPackages(_package.orElse(""))
+				.scan().getClassesWithAnnotation( _annotation.getName() ).getNames();
```
Comment (awittha) [inline src/main/java/com/homeaway/devtools/jenkins/testing/WholeClasspathPipelineExtensionDetector.java]: Extra layer of indentation added here.

```suggestion
		List<String> annotated_classnames = new ClassGraph()
			.enableAnnotationInfo()
			.enableClassInfo()
			.acceptPackages(_package.orElse(""))
			.scan().getClassesWithAnnotation( _annotation.getName() ).getNames();
```

File: src/main/java/com/homeaway/devtools/jenkins/testing/WholeClasspathPipelineExtensionDetector.java [hunk 4]
```diff
@@ -123,11 +131,11 @@
...
-					"Failed to get some classes annotated with [{}] of type [{}] in package [{}]. For detailed error messages, set the system property PipelineExtensionDetector.expandFailures=true. Failures: [{}]",
-					_annotation,
-					_supertype,
-					_package,
-					failures.keySet() );
+						"Failed to get some classes annotated with [{}] of type [{}] in package [{}]. For detailed error messages, set the system property PipelineExtensionDetector.expandFailures=true. Failures: [{}]",
+						_annotation,
+						_supertype,
+						_package,
+						failures.keySet() );
```
Comment (awittha) [inline src/main/java/com/homeaway/devtools/jenkins/testing/WholeClasspathPipelineExtensionDetector.java]: Extra layer of indentation added here.

```suggestion
					"Failed to get some classes annotated with [{}] of type [{}] in package [{}]. For detailed error messages, set the system property PipelineExtensionDetector.expandFailures=true. Failures: [{}]",
					_annotation,
					_supertype,
					_package,
					failures.keySet() );
```

File: src/main/java/com/homeaway/devtools/jenkins/testing/WholeClasspathPipelineExtensionDetector.java [hunk 5]
```diff
@@ -46,7 +46,11 @@
 
 		Set<Class<?>> classes = new HashSet<>();
 
-		List<String> classnames = new FastClasspathScanner(_package.orElse("") ).scan().getNamesOfAllStandardClasses();
+		List<String> classnames = new ClassGraph()
+				.enableAnnotationInfo()
+				.enableClassInfo()
+				.acceptPackages(_package.orElse(""))
+				.scan().getAllStandardClasses().getNames();
```
Comment (awittha) [inline src/main/java/com/homeaway/devtools/jenkins/testing/WholeClasspathPipelineExtensionDetector.java]:
```suggestion
		List<String> classnames = new ClassGraph()
			.enableAnnotationInfo()
			.enableClassInfo()
			.acceptPackages(_package.orElse(""))
			.scan().getAllStandardClasses().getNames();
```

File: CHANGELOG.md [hunk 6]
```diff
@@ -7,7 +7,9 @@ The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/) a
 
 ## Unreleased
 
-* (nothing yet)
+### Updated
+* the deprecated `fast-classpath-scanner` to its latest form `classgraph`.
+* the deprecated `FastClasspathScanner` to its latest form `ClassGraph`.
```
Comment (awittha) [inline CHANGELOG.md:12]:
```suggestion
### Updated

* the deprecated `fast-classpath-scanner` dependency to its latest form `classgraph`.
* the deprecated `FastClasspathScanner` class to its latest form `ClassGraph`.
```

**Review by awittha — CHANGES_REQUESTED** [review]
1. Do existing unit tests cover the correctness of the classpath-scanning code?
2. Is it possible to add a unit test to duplicate the `java.lang.OutOfMemoryError: GC Overhead Limit Exceeded` exception against the original `fast-classpath-scanner`-based codebase, such that the test does _not_ trigger the exception against the `classgraph` codebase?

File: CHANGELOG.md (hunk 6 above)
Comment (awittha) [inline CHANGELOG.md]:
```suggestion
### Updated

* the deprecated `fast-classpath-scanner` dependency to its latest form `classgraph`.
* the deprecated `FastClasspathScanner` class to its latest form `ClassGraph`.
```

**Comment by NingZhuRedsilence** [comment]
> 1. Do existing unit tests cover the correctness of the classpath-scanning code?
> 2. Is it possible to add a unit test to duplicate the `java.lang.OutOfMemoryError: GC Overhead Limit Exceeded` exception against the original `fast-classpath-scanner`-based codebase, such that the test does _not_ trigger the exception against the `classgraph` codebase?

1. I don't think any of the existing *Spec.groovy tests cover the correctness of the classpath-scanning code; there seems to be no Java test in the repo for now.
2. It would be possible to add a unit test as asked for, if we can recreate the state of the JVM where that error happened. But I don't know how at this moment. 

However, since the newly "added" code is just calling a third-party library, I think we / I don't need to write unit test for this third-party code. 

What in `jenkins-spock` are missing unit tests are method `getClassesOfTypeInPackage` and `getClassesWithAnnotationOfTypeInPackage`. I want to assert that adding unit tests to these 2 methods are a different and separate issue from this PR.

**Review by awittha — APPROVED** [review]
